content/posts/Devash_Recommendation_System_Final_Report.md

```diff
@@ -0,0 +1,141 @@
+
+## Why use an unsupervised learning-based model?
+
+The data that was available for us (check out the [database schema](https://github.com/CircuitVerse/CircuitVerse/blob/master/db/schema.rb#L275)) to use didn't have a response variable (typically a X and a Y to train and test the model and improve the accuracy). Therefore, we had to rely on the usage of unsupervised learning-based algorithms to find the intrinsic pattern and hidden structures in the data and present it to the user.
+
+## Model Structure
+
+![Recommendation System Structure](https://i.imgur.com/q5o3GGY.jpg)
+
+## Data Cleaning Module
+
+A SQL query was used to extract the id, author_id, name, description, tag_names, views, and star count which were stored as a JSON file (1,80,000 projects approx). We removed the following:
+
+- HTML tags
+- Non-ASCII characters
+- Punctuations.
+- Replaced NULL values and "Untitled" by empty strings.
+
+  > **Note:** We did not remove punctuations like &, \*, +, etc as they have a meaning in logic circuits.
+
+## Feature Extraction
+
+Feature Extraction is mainly done to extract important parts or features from a huge chunk of items, based on the frequency/count of occurrence and thus reducing the dimensions of the dataset to enhance processing.
+Our recommendation system uses [CountVectorizer](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.CountVectorizer.html) which converts the collection of documents to a matrix of token counts, thus filtering out the important and relevant keywords.
+
+## Topic Modelling using Latent Dirichlet Allocation (LDA)
+
+[LDA](https://scikit-learn.org/stable/modules/generated/sklearn.discriminant_analysis.LinearDiscriminantAnalysis.html) is used to classify text in a document to a particular topic. It builds a topic per document model and words per topic model, modeled as [Dirichlet distributions](https://www.sciencedirect.com/topics/mathematics/dirichlet-distribution).
+
+- Each document is modeled as a multinomial distribution of topics and each topic is modeled as a multinomial distribution of words.
+- LDA assumes that every chunk of text we feed into it will contain words that are somehow related. Therefore choosing the right corpus of data is crucial.
+- It also assumes documents are produced from a mixture of topics. Those topics then generate words based on their probability distribution.
+
+Simply put, LDA creates a distribution of words and packs them together as a topic (the number of topics is chosen by the user). When a new sentence is fed into the model, based on the words present in the new sentence, we get a distribution of the sentence over all the topics. The new dimension of each vector becomes the total number of topics hence drastically reducing the dimensions with better filtering.
+In our recommendation system, the log-likelihood was the least for 10 topics for the given dataset.
```

Comment:
Add bit more stating how each topic is considered as a dimension and the corresponding socres for each topic is used as the new vector representation thus getting a compact form of representation. This is how we do dimensionality reduction which later helps in finding better neighbours in kd trees